ssl/statem/statem_clnt.c

```diff
@@ -1161,7 +1165,248 @@ WORK_STATE ossl_statem_client_post_process_message(SSL_CONNECTION *s,
+    OPENSSL_free(s->ext.ech.innerch);
+    s->ext.ech.innerch = innerch_full;
+    s->ext.ech.innerch_len = innerlen;
+    WPACKET_cleanup(&inner);
+    BUF_MEM_free(inner_mem);
+    inner_mem = NULL;
+# ifdef OSSL_ECH_SUPERVERBOSE
+    /* If tracing, trace out the inner, client random & session id */
+    ech_pbuf("inner CH", s->ext.ech.innerch, s->ext.ech.innerch_len);
+    ech_pbuf("inner, client_random", s->ext.ech.client_random,
+             SSL3_RANDOM_SIZE);
+    ech_pbuf("inner, session_id", s->session->session_id,
+             s->session->session_id_length);
+# endif
+    /* Decode inner so that we can make up encoded inner */
+    if (!PACKET_buf_init(&rpkt, (unsigned char *)s->ext.ech.innerch + 4,
+                         s->ext.ech.innerch_len - 4)) {
+        SSLfatal(s, SSL_AD_INTERNAL_ERROR, ERR_R_INTERNAL_ERROR);
+        goto err;
+    }
+    /*
+     * Parse the full inner CH (usually done on server). This gets us
+     * individually encoded extensions so we can choose to compress
+     * and/or to re-use the same value in outer.
+     */
+    if (!tls_process_client_hello(s, &rpkt)) {
+        SSLfatal(s, SSL_AD_INTERNAL_ERROR, ERR_R_INTERNAL_ERROR);
+        goto err;
+    }
+    /* Make ClientHelloInner and EncodedClientHelloInner as per spec. */
+    if (ech_encode_inner(s) != 1) {
+        SSLfatal(s, SSL_AD_INTERNAL_ERROR, ERR_R_INTERNAL_ERROR);
+        goto err;
+    }
+# ifdef OSSL_ECH_SUPERVERBOSE
+    ech_pbuf("encoded inner CH", s->ext.ech.encoded_innerch,
+             s->ext.ech.encoded_innerch_len);
+# endif
+    /* set deptch for outer CH */
+    s->ext.ech.ch_depth = 0;
+    /*
+     * If we want different key shares for inner and outer, then
+     * zap the one for the inner. The inner key_share is stashed
+     * in s.ext.ech.tmp_pkey already.
+     */
+    if (ech_same_key_share() == 0) {
+        EVP_PKEY_free(s->s3.tmp.pkey);
+        s->s3.tmp.pkey = NULL;
+    }
+    /* Make second call into CH constuction for outer CH. */
+    rv = tls_construct_client_hello_aux(s, pkt);
+    if (rv != 1) {
+        SSLfatal(s, SSL_AD_INTERNAL_ERROR, protverr);
+        goto err;
+    }
+# ifdef OSSL_ECH_SUPERVERBOSE
+    ech_pbuf("outer, client_random", s->s3.client_random, SSL3_RANDOM_SIZE);
+    ech_pbuf("outer, session_id", s->session->session_id,
+             s->session->session_id_length);
+# endif
+    /* Finally, calculate AAD and encrypt using HPKE */
+    if (ech_aad_and_encrypt(s, pkt) != 1) {
+        SSLfatal(s, SSL_AD_INTERNAL_ERROR, ERR_R_INTERNAL_ERROR);
+        goto err;
+    }
+    /* Free up raw exts as needed (happens like this on real server) */
+    if (s->clienthello != NULL
+        && s->clienthello->pre_proc_exts != NULL) {
+        OPENSSL_free(s->clienthello->pre_proc_exts);
+        OPENSSL_free(s->clienthello);
+        s->clienthello = NULL;
+    }
+    return 1;
+err:
+    WPACKET_cleanup(&inner);
+    if (inner_mem != NULL)
+        BUF_MEM_free(inner_mem);
```

Comment:
```suggestion
    if (inner_mem != NULL) {
        WPACKET_cleanup(&inner);
        BUF_MEM_free(inner_mem);
    }
```

otherwise WPACKET_cleanup can be called on uninitialized `inner`.

Reply:
ok

Reply:
actually,  doesn't `inner` need to be `memset` set to zero there - in which case the original is maybe right? Otherwise if `BUF_MEM_new()` works but `BUF_MEM_grow()`  fails we'd still get a non-initialised `inner`

Reply:
Why do you use BUF_MEM_grow() actually? WPACKET will ensure the BUF_MEM is grown appropriately.